### PR TITLE
PWN-7171 - New Swap. Draft for swap state manager

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7171

## Description of Work

- partly done

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapRoutesRefresher.kt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import org.p2p.core.utils.fromLamports
+import org.p2p.core.utils.toLamports
+import org.p2p.wallet.infrastructure.network.provider.TokenKeyProvider
+import org.p2p.wallet.swap.jupiter.repository.model.JupiterSwap
+import org.p2p.wallet.swap.jupiter.repository.model.JupiterSwapRoute
+import org.p2p.wallet.swap.jupiter.repository.routes.JupiterSwapRoutesRepository
+import org.p2p.wallet.swap.jupiter.repository.transaction.JupiterSwapTransactionRepository
+import org.p2p.wallet.utils.toBase58Instance
+
+class SwapRoutesRefresher(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapRoutesRefresher.kt:16]: im not sure about this naming

this is class with reusable logic that can be triggered by different actions

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapRoutesRefresher.kt [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+) {
+    fun refreshRoutes(
+        tokenA: Token.Active,
+        tokenB: Token,
+        amountTokenA: BigDecimal,
+        slippage: Double,
+        activeRouteOrdinal: Int
+    ): Flow<SwapState> {
+        return flow {
+            emit(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapRoutesRefresher.kt:29]: emit loadingRoutes -> render loading Ui
if routes are fetched then emit LoadingTransaction -> render different loading Ui
if transaction is loaded -> emit state with all the data

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateReducer.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package org.p2p.wallet.swap.jupiter.statemanager
+
+import kotlinx.coroutines.flow.Flow
+
+interface SwapStateReducer {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateReducer.kt:5]: each reducer can work with one particular state and create new state depending on action
some states do not react to some actions

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateTokenAZeroReducer.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                    )
+                }
+                is SwapStateAction.TokenAAmountChanged -> {
+                    emitAll(
+                        swapRoutesRefresher.refreshRoutes(
+                            tokenA = oldState.tokenA,
+                            tokenB = oldState.tokenB,
+                            amountTokenA = action.newAmount,
+                            slippage = oldState.slippage,
+                            activeRouteOrdinal = 1
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateTokenAZeroReducer.kt:33]: if current state is TokenAZero, we use first route by default
will extract to const later.

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateTokenAZeroReducer.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                        )
+                    )
+                }
+                is SwapStateAction.TokenAChanged -> {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateTokenAZeroReducer.kt:37]: if TokenAChanged or TokenBChanged or SlippageChanged -> we just change the current TokenAZero state. nothing changes, no requests, we still have 0 in token A input

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateAction.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+sealed class SwapStateAction {
+    data class RouteChanged(val ordinalRouteNumber: Int) : SwapStateAction()
+    data class SlippageChanged(val newSlippageValue: Double) : SwapStateAction()
+
+    data class InitialLoadingFinished(
+        val tokenA: Token.Active,
+        val tokenB: Token
+    ) : SwapStateAction()
+
+    object RefreshRoutes : SwapStateAction()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateAction.kt:15]: every 20 seconds, this action triggers

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateHandler.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package org.p2p.wallet.swap.jupiter.statemanager
+
+import kotlinx.coroutines.flow.Flow
+
+interface SwapStateHandler {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateHandler.kt:5]: like org.p2p.wallet.history.strategy.TransactionParsingContext

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package org.p2p.wallet.swap.jupiter.statemanager
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.collectLatest
+
+class SwapStateManager(private val handlers: Set<SwapStateHandler>) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:8]: the logic is same like in org.p2p.wallet.auth.repository.RestoreUserResultHandler

**Review by NikolaiEmelianov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    private var activeActionHandle: Job? = null
+    private var refreshJob: Job? = null
+
+    fun observe(): StateFlow<SwapState> = state
+
+    fun onNewAction(action: SwapStateAction) {
+        val currentState = state.value
+        val actionHandler = handlers.firstOrNull { it.canHandle(currentState) } ?: return
+
+        activeActionHandle = scope.launch { actionHandler.handleAction(state, action) }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:22]: I can't see where do you cancel this scope when swap fragment is closed

**Review by 1sakov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapState.kt [hunk 10]
```diff
@@ -0,0 +1,43 @@
+package org.p2p.wallet.swap.jupiter.statemanager
+
+import java.math.BigDecimal
+import org.p2p.core.token.Token
+import org.p2p.solanaj.utils.crypto.Base64String
+import org.p2p.wallet.swap.jupiter.repository.model.JupiterSwapRoute
+
+sealed interface SwapState {
+    object InitialLoading : SwapState
+    data class TokenAZero(
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapState.kt:10]: add newline between classes

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateAction.kt [hunk 11]
```diff
@@ -0,0 +1,25 @@
+package org.p2p.wallet.swap.jupiter.statemanager
+
+import java.math.BigDecimal
+import org.p2p.core.token.Token
+
+sealed class SwapStateAction {
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateAction.kt:6]: can you please add newline between classes

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateHandler.kt [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package org.p2p.wallet.swap.jupiter.statemanager
+
+import kotlinx.coroutines.flow.MutableStateFlow
+
+interface SwapStateHandler {
+    fun canHandle(oldState: SwapState): Boolean
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateHandler.kt:6]: why old state ?

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt [hunk 13]
```diff
@@ -0,0 +1,24 @@
+package org.p2p.wallet.swap.jupiter.statemanager
+
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+class SwapStateManager(
+    private val handlers: Set<SwapStateHandler>,
+    private val scope: SwapCoroutineScope
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:10]: for what we create swapScope ? which difference between this and our IO scope ?

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+) {
+    private val state = MutableStateFlow<SwapState>(SwapState.InitialLoading)
+    private var activeActionHandle: Job? = null
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:13]: activeActionHandeJob

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateTokenAZeroHandler.kt [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+                state.value = oldState.copy(slippage = action.newSlippageValue)
+            }
+            is SwapStateAction.TokenAAmountChanged -> {
+                swapRoutesRefresher.refreshRoutes(
+                    state = state,
+                    tokenA = oldState.tokenA,
+                    tokenB = oldState.tokenB,
+                    amountTokenA = action.newAmount,
+                    slippage = oldState.slippage,
+                    activeRouteOrdinal = 1
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateTokenAZeroHandler.kt:26]: move number to const

**Review by Davran331122 — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt (hunk 13 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:10]: scope is active while Swap feature is not closed
it is used to refresh Swap routes between different screens on Swap

**Review by Davran331122 — APPROVED** [review]

**Pull request merged**